hoptimator-util/src/main/java/com/linkedin/hoptimator/util/planner/PipelineRel.java

```diff
@@ -36,36 +44,40 @@
 public interface PipelineRel extends RelNode {
 
   Convention CONVENTION = new Convention.Impl("PIPELINE", PipelineRel.class);
+  String KEY_OPTION = "keys";
+  String KEY_PREFIX = "KEY_";
 
   void implement(Implementor implementor) throws SQLException;
 
   /** Implements a deployable Pipeline. */
   class Implementor {
     private final Map<Source, RelDataType> sources = new LinkedHashMap<>();
     private RelNode query;
+    private ImmutableList<Pair<Integer, String>> targetFields;
     private String sinkDatabase = "pipeline";
     private List<String> sinkPath = Arrays.asList(new String[]{"PIPELINE", "SINK"});
     private RelDataType sinkRowType = null;
     private Map<String, String> sinkOptions = Collections.emptyMap();
 
-    public void visit(RelNode node) throws SQLException {
-      if (query == null) {
-        query = node;
+    public void visit(RelNode node, ImmutableList<Pair<Integer, String>> targetFields) throws SQLException {
```

Comment:
targetFields probably makes more sense as a constructor parameter, since it doesn't appear to change for the life of the Implementor object. Also it is weird to see the visit() method break from the "visitor pattern". Probably doesn't matter for us, but sometimes visit() is invoked via reflection, so the signature probably shouldn't change.